File: python3/remove_duplicates_from_sorted_list_II.py
from model import ListNode


"""
Space   : O(1)
Time    : O(n)
"""

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        
        ph = head
        ans = ListNode(0)
        pa = ans
        
        while ph:
            if not ph.next:
                pa.next = ph
                break
            elif ph.val == ph.next.val:
                rem = ph.val
                while ph.val == rem:
                    ph = ph.next
                    if not ph:
                        break
            else:
                pa.next = ListNode(ph.val)
                pa = pa.next
                ph = ph.next

        return ans.next

File: golang/remove_duplicates_from_sorted_list_II.go
package golang

import (
	"golang/model"
)

/*
Space	: O(1)
Time	: O(n)
*/

func deleteDuplicates(head *model.ListNode) *model.ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	ph := head
	ans := &model.ListNode{
		Val: 0,
	}
	pans := ans

	for ph != nil {
		if ph.Next == nil {
			pans.Next = ph
			break
		} else if ph.Val == ph.Next.Val {
			rem := ph.Val
			for ph.Val == rem {
				ph = ph.Next
				if ph == nil {
					break
				}
			}
		} else {
			pans.Next = &model.ListNode{
				Val: ph.Val,
			}
			pans = pans.Next
			ph = ph.Next
		}
	}

	return ans.Next
}
